tartufo/config.py

```diff
@@ -255,20 +260,47 @@ def compile_rule(pattern: str) -> Rule:
         path, pattern = pattern.split("::", 1)
     except ValueError:  # Raised when the split separator is not found
         path = ".*"
-    return Rule(name=None, pattern=re.compile(pattern), path_pattern=re.compile(path))
+    return Rule(
+        name=None,
+        pattern=re.compile(pattern),
+        path_pattern=re.compile(path),
+        re_match_type="match",
+    )
 
 
-def compile_rules(patterns: Iterable[str]) -> List[Rule]:
+def compile_rules(patterns: Iterable[Union[str, Dict[str, str]]]) -> List[Rule]:
     """Take a list of regex string with paths and compile them into a List of Rule.
 
     Any line starting with `#` will be ignored.
 
     :param patterns: The list of patterns to be compiled
     :return: List of Rule objects
     """
-    stripped = (p.strip() for p in patterns)
-    return [
-        compile_rule(pattern)
-        for pattern in stripped
-        if pattern and not pattern.startswith("#")
-    ]
+    try:
+        return list(
+            {
+                Rule(
+                    name=pattern.get("reason", None),  # type: ignore[union-attr]
+                    pattern=re.compile(pattern["pattern"]),  # type: ignore[index]
+                    path_pattern=re.compile(pattern.get("files", ".*")),  # type: ignore[union-attr]
```

Comment:
```suggestion
                    path_pattern=re.compile(pattern.get("files", "")),  # type: ignore[union-attr]
```
For the same reasons as above